### DISCUSSION TITLE
Support for time-based animations.

### DISCUSSION THREAD

**Discussion by AnBenLa** [syoyo/tinyusdz#104] [General]
Hello,
as I am new to this repository, I was wondering if it already supports the creation of USD animations.
Specifically, I'm interested in using keyframe transform data to create a USD file containing the animation.
Is this already supported in the current state of the project?

**Comment by syoyo** [comment]
Yes, timesampled attribute(animation) is supported in TinyUSDZ.

If you want to create a USD with animation attributes, see `api-tutorial` example

https://github.com/syoyo/tinyusdz/blob/469dbed09f43d5bf8a3396d2fa61c5ddae4aaea5/examples/api_tutorial/api-tutorial-main.cc#L125
(for xform ops)

and `Animatable` class for generic animatable Attribute(e.g. SkelAnimation.rotations https://github.com/syoyo/tinyusdz/blob/469dbed09f43d5bf8a3396d2fa61c5ddae4aaea5/src/usdSkel.hh#L250 )

https://github.com/syoyo/tinyusdz/blob/469dbed09f43d5bf8a3396d2fa61c5ddae4aaea5/src/prim-types.hh#L1106

As you may know, USD only supports Linear or Held(something like `nearest_neighbor`) interpolation, so need to convert slerp, cubic, etc keyframe to linear by oversampling keyframe samples. Codes in https://github.com/google/usd_from_gltf would help converting keyframe data for USD linear timesampled data.

**Reply by AnBenLa** [reply]
Thank you for the extensive reply and the note about oversampling keyframe samples!
For my use case, I settled for using time samples for the translation, rotation and scaling operations.
```c++
tinyusdz::XformOp translateOp;
translateOp.op_type = tinyusdz::XformOp::OpType::Translate;
translateOp.set_timesample(t0, pos0);
translateOp.set_timesample(t1, pos1);
```

**Reply by syoyo** [reply]
👍
